## Preprocessing

In [2]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("Resources/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(["EIN", "NAME"], axis=1)
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [4]:
# Determine the number of unique values in each column.
for column in application_df.columns:

    line_length = 32
    row_string = ""
    number = len(application_df[column].unique())

    number_length = len(str(number))
    column_length = len(column)
    spacing_length = line_length - (number_length + column_length)

    row_string += column
    for i in range(spacing_length):
        row_string += " "

    
    row_string += str(number)


    print(row_string)

APPLICATION_TYPE              17
AFFILIATION                    6
CLASSIFICATION                71
USE_CASE                       5
ORGANIZATION                   4
STATUS                         2
INCOME_AMT                     9
SPECIAL_CONSIDERATIONS         2
ASK_AMT                     8747
IS_SUCCESSFUL                  2


In [5]:
application_df["APPLICATION_TYPE"].value_counts()


T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T14        3
T25        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [6]:
apps = application_df["APPLICATION_TYPE"].value_counts()

application_types_to_replace = []
count_cutoff = 500

for index in application_df["APPLICATION_TYPE"].value_counts().index:
    if application_df["APPLICATION_TYPE"].value_counts()[index] < count_cutoff:
        application_types_to_replace.append(index)
print(application_types_to_replace)


['T9', 'T13', 'T12', 'T2', 'T14', 'T25', 'T29', 'T15', 'T17']


In [7]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`

apps = application_df["APPLICATION_TYPE"].value_counts()

application_types_to_replace = []
count_cutoff = 500

for index in application_df["APPLICATION_TYPE"].value_counts().index:
    if application_df["APPLICATION_TYPE"].value_counts()[index] < count_cutoff:
        application_types_to_replace.append(index)


# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [8]:
# Look at CLASSIFICATION value counts for binning
application_df["CLASSIFICATION"].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4200        1
C1283        1
C1570        1
C2500        1
C8210        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [9]:
# You may find it helpful to look at CLASSIFICATION value counts >1
application_df["CLASSIFICATION"].value_counts().loc[application_df["CLASSIFICATION"].value_counts() > 1]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C2300       32
C7200       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1278       10
C1238       10
C1235        9
C1237        9
C7210        7
C1720        6
C4100        6
C2400        6
C1257        5
C1600        5
C2710        3
C1260        3
C0           3
C1267        2
C1234        2
C1256        2
C3200        2
C1246        2
Name: CLASSIFICATION, dtype: int64

In [10]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`

classifications_to_replace = []
count_cutoff = 1500

for index in application_df["CLASSIFICATION"].value_counts().index:
    if application_df["CLASSIFICATION"].value_counts()[index] < count_cutoff:
        classifications_to_replace.append(index)


# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [11]:
# Convert categorical data to numeric with `pd.get_dummies`
application_numberic_df = pd.get_dummies(application_df)
application_numberic_df.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                       0   
1       1   108590              1                       0   
2       1     5000              0                       0   
3       1     6692              1                       0   
4       1   142590              1                       0   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                     1                     0                    0   
1                     0                     0                    1   
2                     0                     0                    0   
3                     0                     0                    1   
4                     0                     0                    1   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                    0                    0                    0  ...   
1                    0                    0                    0  ...   
2                    0                    1                    0  ...   
3                    0                    0                    0  ...   
4                    0                    0                    0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                  0                       0                         0   
1                  1                       0                         0   
2                  0                       0                         0   
3                  0                       1                         0   
4                  0                       0                         1   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                   0                 0                       0   
1                   0                 0                       0   
2                   0                 0                       0   
3                   0                 0                       0   
4                   0                 0                       0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0                0                  0                         1   
1                0                  0                         1   
2                0                  0                         1   
3                0                  0                         1   
4                0                  0                         1   

   SPECIAL_CONSIDERATIONS_Y  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  

[5 rows x 44 columns]

In [12]:
# Split our preprocessed data into our features and target arrays
target = application_numberic_df["IS_SUCCESSFUL"]
features = application_numberic_df.drop(["IS_SUCCESSFUL"], axis=1)

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(features, target, random_state=10)

In [13]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [14]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=80, activation="relu", input_dim=43))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=30, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 80)                3520      
_________________________________________________________________
dense_1 (Dense)              (None, 30)                2430      
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 31        
Total params: 5,981
Trainable params: 5,981
Non-trainable params: 0
_________________________________________________________________


In [15]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [16]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=20)

Epoch 1/20
804/804 [==============================] - 1s 904us/step - loss: 0.5686 - accuracy: 0.7231
Epoch 2/20
804/804 [==============================] - 1s 865us/step - loss: 0.5572 - accuracy: 0.7291
Epoch 3/20
804/804 [==============================] - 1s 769us/step - loss: 0.5528 - accuracy: 0.7325
Epoch 4/20
804/804 [==============================] - 1s 857us/step - loss: 0.5523 - accuracy: 0.7323
Epoch 5/20
804/804 [==============================] - 1s 744us/step - loss: 0.5510 - accuracy: 0.7304
Epoch 6/20
804/804 [==============================] - 1s 804us/step - loss: 0.5493 - accuracy: 0.7339
Epoch 7/20
804/804 [==============================] - 1s 822us/step - loss: 0.5491 - accuracy: 0.7326
Epoch 8/20
804/804 [==============================] - 1s 758us/step - loss: 0.5475 - accuracy: 0.7350
Epoch 9/20
804/804 [==============================] - 1s 756us/step - loss: 0.5469 - accuracy: 0.7340
Epoch 10/20
804/804 [==============================] - 1s 763us/step - loss: 0.545

In [17]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5538 - accuracy: 0.7244
Loss: 0.5538440346717834, Accuracy: 0.7244315147399902


In [18]:
# Export our model to HDF5 file
nn.save_weights("application.hdf5")

## Keras Tuner Search

The model that I made in the last section only has a 72.4% accuracy, so I tried to use keras tuner to find a better set of parameters to hopefully get it above 75%. It didn't really work out, getting only 0.4% more accurate after running through 60 different trials

In [19]:
import keras_tuner
from tensorflow import keras

In [20]:
def create_model(hp):
    nn_model = tf.keras.models.Sequential()

    # Allow kerastuner to decide which activation function to use in hidden layers
    activation = hp.Choice('activation',['relu','tanh','sigmoid'])
    
    # Allow kerastuner to decide number of neurons in first layer
    nn_model.add(tf.keras.layers.Dense(
        units=hp.Int('first_units',
            min_value=1,
            max_value=80,
            step=2
        ),
        activation=activation,
        input_dim=43
        ))

    # Allow kerastuner to decide number of hidden layers and neurons in hidden layers
    for i in range(hp.Int('num_layers', 1, 6)):
        nn_model.add(tf.keras.layers.Dense(
            units=hp.Int('units_' + str(i),
                min_value=1,
                max_value=80,
                step=2),
            activation=activation))
    
    nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

    # Compile the model
    nn_model.compile(loss="binary_crossentropy", optimizer='adam', metrics=["accuracy"])
    
    return nn_model

In [21]:
tuner = keras_tuner.RandomSearch(
    create_model,
    objective="val_accuracy",
    max_trials=5)


INFO:tensorflow:Reloading Oracle from existing project .\untitled_project\oracle.json
INFO:tensorflow:Reloading Tuner from .\untitled_project\tuner0.json


In [26]:
tuner.search(X_train_scaled, y_train, epochs=20, validation_data=(X_test_scaled, y_test))

Trial 60 Complete [00h 00m 19s]
val_accuracy: 0.7253644466400146

Best val_accuracy So Far: 0.7285131216049194
Total elapsed time: 00h 07m 34s
INFO:tensorflow:Oracle triggered exit


In [27]:
best_hyper = tuner.get_best_hyperparameters(1)[0]
best_hyper.values

{'activation': 'tanh',
 'first_units': 49,
 'num_layers': 6,
 'units_0': 13,
 'units_1': 19,
 'units_2': 33,
 'units_3': 37,
 'units_4': 67,
 'tuner/epochs': 20,
 'tuner/initial_epoch': 0,
 'tuner/bracket': 0,
 'tuner/round': 0,
 'units_5': 1}

In [28]:
best_model = tuner.get_best_models(1)[0]
model_loss, model_accuracy = best_model.evaluate(X_test_scaled, y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5568 - accuracy: 0.7285
Loss: 0.5567918419837952, Accuracy: 0.7285131216049194


In [29]:
best_model.save_weights("application_tuner_search.hdf5")

## Report on teh Neural Network Model

The purpose of this neural net is to predict the outcome of applicants for Alphabet Soup funding. 

By taking an input of ### together with the outcome 

The initial neural net was made of three layers; 2 hidden layers with 80 and 60 neurons and a final output layer with a single neuron indicating a binary value for success/failure. Performance only reached 72.4% accuracy, falling short of the target performance . Hoping to improve the model I used Keras Tuner to search for optimal neural nets with between 1 and 6 layers and 1 to 80 neurons, this netted only a very marginal improvement with an accuracy of 72.85% 

Nearly all of the models tried with this approach got results in the 69%-72% range, indicating that this may be a limitation with the data itself, more accurate models may simply not be possible.